torchgeo/datasets/splits.py

```diff
@@ -0,0 +1,359 @@
+        lengths[idx_to_add_at] += 1
+    return lengths
+
+
+def random_bbox_assignment(
+    dataset: GeoDataset,
+    lengths: Sequence[Union[int, float]],
+    generator: Optional[Generator] = default_generator,
+) -> List[GeoDataset]:
+    """Split a GeoDataset randomly assigning its index's BoundingBoxes.
+
+    Args:
+        dataset: dataset to be split
+        lengths: lengths or fractions of splits to be produced
+        generator: (optional) generator used for the random permutation
+
+    Returns
+        A list of the subset datasets.
+
+    .. versionadded:: 0.5
+    """
+    if not (sum(lengths) == 1 or sum(lengths) == len(dataset)):
+        raise ValueError(
+            "Sum of input lengths must equal 1 or the length of dataset's index."
+        )
+
+    if any(n <= 0 for n in lengths):
+        raise ValueError("All items in input lengths must be greater than 0.")
+
+    if sum(lengths) == 1:
+        lengths = _fractions_to_lengths(lengths, len(dataset))
+
+    hits = list(dataset.index.intersection(dataset.index.bounds, objects=True))
+
+    hits = [
+        hits[i]
+        for i in randperm(sum(lengths), generator=generator)  # type: ignore[arg-type]
+    ]
+
+    new_indexes = [
+        Index(interleaved=False, properties=Property(dimension=3)) for _ in lengths
+    ]
+
+    for i, length in enumerate(lengths):
+        for j in range(length):  # type: ignore[arg-type]
+            hit = hits.pop()
+            new_indexes[i].insert(j, hit.bounds, hit.object)
+
+    new_datasets = []
+    for index in new_indexes:
+        ds = deepcopy(dataset)
+        ds.index = index
+        new_datasets.append(ds)
+
+    return new_datasets
+
+
+def random_bbox_splitting(
+    dataset: GeoDataset,
+    fractions: Sequence[float],
+    generator: Optional[Generator] = default_generator,
+) -> List[GeoDataset]:
+    """Split a GeoDataset randomly splitting its index's BoundingBoxes.
+
+    This function will go through each BoundingBox in the GeoDataset's index and
+    split it in a random direction.
+
+    Args:
+        dataset: dataset to be split
+        fractions: fractions of splits to be produced
+        generator: (optional) generator used for the random permutation
+
+    Returns
+        A list of the subset datasets.
+
+    .. versionadded:: 0.5
+    """
+    if sum(fractions) != 1:
+        raise ValueError("Sum of input fractions must equal 1.")
+
+    if any(n <= 0 for n in fractions):
+        raise ValueError("All items in input fractions must be greater than 0.")
+
+    new_indexes = [
+        Index(interleaved=False, properties=Property(dimension=3)) for _ in fractions
+    ]
+
+    for i, hit in enumerate(
+        dataset.index.intersection(dataset.index.bounds, objects=True)
+    ):
+        box = BoundingBox(*hit.bounds)
+        fraction_left = 1.0
+
+        horizontal, flip = randint(0, 2, (2,), generator=generator)
+        for j, frac in enumerate(fractions):
```

Comment:
How does this work when len(fractions) > 2? Do we just choose a random direction (horizontal, vertical) and then split a square into rectangles with a certain fractional width?

Reply:
For len(fractions) > 2 it first splits the bbox using the first fraction, where one resulting bbox's area will correspond to that fraction and the second will have to be splitted again. So we now split that second bbox using the second fraction, adjusted to the remaining area. And we keep repeating until done.

Not sure if it's clear enough.

Reply:
Might be useful to add more comments to the code. Some of the logic is difficult to follow, not because it's overly complicated but because there's no easier way to do it.